### ISSUE TITLE
Cannot Keyboard focus into TreeView Component

### ISSUE THREAD

**Issue by garyb1** [dgreene1/react-accessible-treeview#48]
Hi I was wondering if someone can help me please?

In the [w3 APG TreeView Pattern](https://www.w3.org/WAI/ARIA/apg/patterns/treeview/) it says:

> When a single-select tree receives focus:
> If none of the nodes are selected before the tree receives focus, focus is set on the first node.
> If a node is selected before the tree receives focus, focus is set on the selected node.

If I tab to the first tree item of the tree view I can see the tabindex attribute is set to 0. This is good. 
If I tab again and it takes me away from the tree view component to the next interactive element. This is expected. 

What I noticed is that when navigating away, the component sets the first tree item to tabindex="-1" which means if I press shift and tab I will no longer be able to navigate to the treeview first item (or any item). 

The first node should be selected when I tab back to the component but this isn't the case. 

The same thing happens when you select a node and tab away and tab back more than once. 

[Codesandbox Link](https://codesandbox.io/s/quiet-dawn-k6uglp?file=/src/App.js)

**Comment by dgreene1** [comment]
@mellis481 perhaps a member of your team can take on reproducing this.

@garyb1 our team can’t commit to a fix at this time. We will update this ticket if we can. But we are open to contributions if you would like to submit a PR. Are you interested?

We’ve yet to find a more accessible treeview available, so that’s why we recently took over as maintainers. It’s not to say that there aren’t improvements that can be made (like this finding)— but it is to say that if you’re willing to improve this library by making a PR, it would be very helpful to all React users who need a tree view.

**Comment by garyb1** [comment]
Hey @dgreene1,
Sure thing. I will take a look tomorrow morning (GMT+1)

**Comment by mellis481** [comment]
@garyb1 Thanks for reporting this issue. I was able to reproduce it in your example and confirm it's a valid bug.  As @dgreene1 said, if you are able to submit a PR to address this, we'd be happy to review it.

**Comment by garyb1** [comment]
No problem @mellis481 , I will try my best. 

Unrelated to the above:
Trying to install the project here using node v18.10.0 and getting lots of issues with unmet peer dependencies. Looking to try and resolve it but would love to know if anyone else has that issue?

**Comment by garyb1** [comment]
@mellis481  @dgreene1 

I'm having lots of issues with installing the app such as peer dependency issues, rollup issues with output directory error, and even typescript. Is there any chance you can add the installation steps you use to the project? Node version? etc. 

Or even as part of a contributing guide.

**Comment by dgreene1** [comment]
Hi @garyb1, we’re on Node LTS (so 16.17.1).

Do you use NVM? If so, switch to your LTS installation and then install.

That should solve everything since Github Actions is currently building it fine in 16.15           https://github.com/dgreene1/react-accessible-treeview/blob/8bbe116f9e8bdd65715d1e253de59c231dbec722/.github/workflows/pr-gate.yml#L16 and it works for me on 16.17.1

**Comment by garyb1** [comment]
@dgreene1  Thanks for getting back to me. Yeah I used Volta and installed LTS. I'll try again on my personal machine later.

**Comment by kpustakhod** [comment]
@garyb1  with node 16.13.0 the installation steps are:
1. `npm i` in root folder
2. `npm i` in website folder
3. `npm run build-storybook` in root folder
4. `npm run storybook` in root folder

If you need to make any changes in source code and want them to appear in storybook you would need to:
1. `npm link` in root folder
2. `npm link react-accessible-treeview` in website folder
Additionally run `npm run build` after each change in root folder.

**Comment by garyb1** [comment]
@kpustakhod That worked perfectly with those instructions. It was the node version that fixed it for me with 16.13.0.
App is running on my machine now. Thanks very much.

**Comment by mellis481** [comment]
@garyb1 Good to hear. We will work to get this library to play nice with Node LTS.

**Comment by garyb1** [comment]
@mellis481  @dgreene1 @kpustakhod 

The above is not a bug with the component. It is actually coming from the docs DirectoryView example which I was using in my code sandbox example above.

What happened is we had an onBlur event handler which dispatched the deselect action. 

```
onBlur={({ treeState, dispatch }) => {
  dispatch({
    type: "DESELECT",
    id: Array.from(treeState.selectedIds)[0],
  });
}}
```

This action checks the  _tabbableId_ field with `action.keepFocus ? state.tabbableId : action.id`. We weren't passing keepFocus in the action in the example so then _tabbledid_ would get set to undefined whenever the focus went away from the component (onBlur).

I removed the onBlur event handler and verified everything worked as expected. 

It seems to work when we add the keepFocus field to the action.

```
onBlur={({ treeState, dispatch }) => {
dispatch({
  type: "DESELECT",
  keepFocus: true,
  id: Array.from(treeState.selectedIds)[0],
});
}}
```

**Comment by garyb1** [comment]
I have one concern though and I would love to know if this is your expected behaviour (see video attached).

In the APG Tree View Design Pattern notes it says:

For a vertically oriented tree:

> When a single-select tree receives focus:
> If none of the nodes are selected before the tree receives focus, focus is set on the first node.
> If a node is selected before the tree receives focus, focus is set on the selected node.

When a tree item receives focus using the keyboard it has a tabindex value of 0. Then every other field is set to tabindex -1. This is using roving tabindex I guess.

If I tab away from the tree view, the currently focused item will have a tabindex of 0. When I tab back into the tree view component that previous item will regain keyboard focus because it still has a tabindex of 0. 

But should it? That item is not in a  "selected" state.  See the video, I did not select the item but it regained focus. Shouldn't item 0 (from default state) regain focus? or is this a behaviour you desire?

https://user-images.githubusercontent.com/12486130/194406964-45824959-b755-4d22-b174-53308999547a.mov

**Comment by dgreene1** [comment]
@garyb1 if you could, please make a ticket for the other issue. I’d like to close this one since you clarified that it’s not an issue.

**Comment by mellis481** [comment]
@garyb1 The tabindex-related issue you originally posted is definitely something that should be logged so please make sure you create an issue for that (unless you were going to create a PR directly 😃). Thanks!

**Comment by garyb1** [comment]
> @garyb1 The tabindex-related issue you originally posted is definitely something that should be logged so please make sure you create an issue for that (unless you were going to create a PR directly 😃). Thanks!

Will do @mellis481 .. I logged the issue but will create a PR for it tomorrow. Thanks.

**Comment by mellis481** [comment]
@garyb1 Ah OK. I thought the issue you were reporting in [this](https://github.com/dgreene1/react-accessible-treeview/issues/48#issuecomment-1270610724) comment was different than what you reporting in your original comment in this issue.

**Comment by garyb1** [comment]
@mellis481 sorry yes I was confused. My bad. I'll create a ticket for the keepFocus docs issue also. I'll do that tomorrow.
